src/cursor/cur_version.c

```diff
@@ -0,0 +1,470 @@
+                durable_stop_ts = version_cursor->upd_durable_stop_ts;
+                stop_ts = version_cursor->upd_stop_ts;
+                stop_txn = version_cursor->upd_stop_txnid;
+            } else {
+                durable_stop_ts = vpack->tw.durable_stop_ts;
+                stop_ts = vpack->tw.stop_ts;
+                stop_txn = vpack->tw.stop_txn;
+            }
+
+            __wt_cursor_set_key(cursor, vpack->tw.start_txn, vpack->tw.start_ts,
+              vpack->tw.durable_start_ts, stop_txn, stop_ts, durable_stop_ts, WT_UPDATE_STANDARD,
+              tw.prepare, 0, WT_VERSION_DISK_IMAGE);
+            cursor->value.data = vpack->data;
+            cursor->value.size = vpack->size;
+            F_SET(cursor, WT_CURSTD_VALUE_INT);
+
+            upd_found = true;
+            F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+        }
+    }
+
+    if (!upd_found && !F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /*
+         * If the history store cursor is not yet positioned, then we are traversing the history
+         * store versions for the first time.
+         */
+        if (!F_ISSET(hs_cursor, WT_CURSTD_KEY_INT)) {
+            hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, cursor->key, WT_TS_MAX, UINT64_MAX);
+            WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, hs_cursor), true);
+        } else
+            WT_ERR_NOTFOUND_OK(ret = hs_cursor->prev(hs_cursor), true);
+
+        /*
+         * If there are no history store records for the given key or if we have iterated through
+         * all the records already, we have exhausted the history store.
+         */
+        if (ret == 0) {
+            WT_TIME_WINDOW_INIT(&tw);
+            __wt_hs_upd_time_window(hs_cursor, &twp);
+            WT_TIME_WINDOW_COPY(twp, &tw);
+            WT_ERR(hs_cursor->get_value(
+              hs_cursor, &tw.stop_ts, &tw.durable_start_ts, &hs_upd_type, &hs_value));
+            __wt_cursor_set_key(cursor, tw.start_txn, tw.start_ts, tw.durable_start_ts, tw.stop_txn,
+              tw.stop_ts, tw.durable_stop_ts, hs_upd_type, 0, 0, WT_VERSION_HISTORY_STORE);
+
+            /*
+             * Reconstruct the history store value if needed. Since we save the value inside the
+             * version cursor every time we traverse a version, we can simply apply the modify onto
+             * the latest value.
+             */
+            if (hs_upd_type == WT_UPDATE_MODIFY) {
+                WT_ERR(__wt_modify_apply_item(
+                  session, table_cursor->value_format, &cursor->value, hs_value.data));
+            } else {
+                WT_ASSERT(session, hs_upd_type == WT_UPDATE_STANDARD);
+                cursor->value.data = hs_value.data;
+                cursor->value.size = hs_value.size;
+            }
+            F_SET(cursor, WT_CURSTD_VALUE_INT);
+            upd_found = true;
+        } else
+            F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
```

Comment:
No need to set this here.

Reply:
I don't really see a point of this `WT_VERSION_CUR_HS_EXAUSTED` constant now, do you think we should get rid of this completely?

Reply:
You still set it when there is no on disk value to indicate no need to try hs.